### PR TITLE
Reapply netlink retries

### PR DESCRIPTION
**Pull request by hefroy**
Reapplying commit a268e2d68d30ea9b09834dbaca60530b4314774c from @phiwer with #283 which was removed by commit 826ebb8d352245a36ecaec32b6af61e7abf4696e from @DiogoPedrozza.

Additionally:
1. limit the number of retries to three, since otherwise a persistent error would cause infinite retries and high CPU load.
2. Improve the logging, to only output the error message after the retries are exhausted and fix the error message output, since `strerror` takes a positive `errno`.

### CODE REVIEW

File: implementation/endpoints/include/netlink_connector.hpp [hunk 1]
```diff
@@ -161,9 +161,9 @@ class netlink_connector : public std::enable_shared_from_this<netlink_connector>
     bool has_address(struct ifaddrmsg * ifa_struct,
             size_t length,
             const unsigned int address);
-    void send_ifa_request();
-    void send_ifi_request();
-    void send_rt_request();
+    void send_ifa_request(unsigned int retry = 0);
+    void send_ifi_request(unsigned int retry = 0);
+    void send_rt_request(unsigned int retry = 0);
```
Comment (anaritarodrigues) [inline implementation/endpoints/include/netlink_connector.hpp:166]: just to be coherent with the rest of the code.
```suggestion
    void send_ifa_request(std::uint32_t retry = 0);
    void send_ifi_request(std::uint32_t retry = 0);
    void send_rt_request(std::uint32_t retry = 0);
```

File: implementation/endpoints/src/netlink_connector.cpp [hunk 2]
```diff
@@ -17,6 +17,15 @@
...
 namespace vsomeip_v3 {
 
+namespace {
+    const std::uint32_t max_retries = 3;
+    const std::uint32_t retry_bit_shift = 8;
+    const std::uint32_t request_sequence_bitmask = 0xFF;
+    const std::uint32_t ifa_request_sequence = 1;
+    const std::uint32_t ifi_request_sequence = 2;
+    const std::uint32_t rt_request_sequence = 3;
+}
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:27]: Move these variables to netlink_connector.hpp as a static const member.

File: implementation/endpoints/src/netlink_connector.cpp [hunk 3]
```diff
@@ -180,9 +189,37 @@ void netlink_connector::receive_cbk(boost::system::error_code const &_error,
...
                     struct nlmsgerr *errmsg = (nlmsgerr *)NLMSG_DATA(nlh);
-                    VSOMEIP_ERROR << "netlink_connector::receive_cbk received "
-                        "error message: " << std::dec << nlh->nlmsg_type
-                        << " seq " << errmsg->msg.nlmsg_seq;
+                    if (errmsg->error == 0) {
+                        // Ack from netlink
+                        break;
+                    }
+
+                    unsigned int retry = errmsg->msg.nlmsg_seq >> retry_bit_shift;
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:197]: just to be coherent with the rest of the code.
```suggestion
                    std::uint32_t retry = errmsg->msg.nlmsg_seq >> retry_bit_shift;
```

File: implementation/endpoints/src/netlink_connector.cpp [hunk 4]
```diff
@@ -180,9 +189,37 @@ void netlink_connector::receive_cbk(boost::system::error_code const &_error,
...
+                    if (retry >= max_retries) {
+                        VSOMEIP_ERROR << "netlink_connector::receive_cbk received "
+                            "error message: " << strerror(-errmsg->error)
+                            << " type " << std::dec << errmsg->msg.nlmsg_type
+                            << " seq " << errmsg->msg.nlmsg_seq;
+                        break;
+                    }
+
+                    unsigned int request_sequence = errmsg->msg.nlmsg_seq & request_sequence_bitmask;
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:206]: just to be coherent with the rest of the code.
```suggestion
                   std::uint32_t request_sequence = errmsg->msg.nlmsg_seq & request_sequence_bitmask;
```

File: implementation/endpoints/src/netlink_connector.cpp [hunk 5]
```diff
@@ -240,7 +277,7 @@ void netlink_connector::send_cbk(boost::system::error_code const &_error, std::s
     }
 }
 
-void netlink_connector::send_ifa_request() {
+void netlink_connector::send_ifa_request(unsigned int retry) {
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:280]:
```suggestion
void netlink_connector::send_ifa_request(std::uint32_t int retry) {
```

File: implementation/endpoints/src/netlink_connector.cpp [hunk 6]
```diff
@@ -268,7 +305,7 @@ void netlink_connector::send_ifa_request() {
     );
 }
 
-void netlink_connector::send_ifi_request() {
+void netlink_connector::send_ifi_request(unsigned int retry) {
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:308]:
```suggestion
void netlink_connector::send_ifi_request(std::uint32_t retry) {
```

File: implementation/endpoints/src/netlink_connector.cpp [hunk 7]
```diff
@@ -295,7 +332,7 @@ void netlink_connector::send_ifi_request() {
     }
 }
 
-void netlink_connector::send_rt_request() {
+void netlink_connector::send_rt_request(unsigned int retry) {
```
Comment (anaritarodrigues) [inline implementation/endpoints/src/netlink_connector.cpp:335]:
```suggestion
void netlink_connector::send_rt_request(std::uint32_t retry) {
```

**Review by anaritarodrigues — CHANGES_REQUESTED** [review]

File: implementation/endpoints/src/netlink_connector.cpp (hunk 2 above)
Comment (hefroy) [reply implementation/endpoints/src/netlink_connector.cpp]: Done :heavy_check_mark:

**Review by anaritarodrigues — CHANGES_REQUESTED** [review]
Please apply the following patch 
[0001-creted-method.zip](https://github.com/COVESA/vsomeip/files/13977581/0001-creted-method.zip)

**Comment by hefroy** [comment]
> Please apply the following patch [0001-creted-method.zip](https://github.com/COVESA/vsomeip/files/13977581/0001-creted-method.zip)

Applied :heavy_check_mark:

**Review by DiogoPedrozza — APPROVED** [review]
LGTM

**Review by anaritarodrigues — APPROVED** [review]
Thank you for the input.

**Comment by DiogoPedrozza** [comment]
@phiwer and @hefroy thank you for your support!

**Pull request merged**